graphql_compiler/schema_transformation/split_query.py

```diff
@@ -218,36 +343,38 @@ def _split_query_one_level(
 
 
 def _split_query_ast_one_level_recursive(
-    query_node, ast, type_info, edge_to_stitch_fields, name_assigner
-):
+    query_node: SubQueryNode,
+    ast: AstType,
+    type_info: TypeInfo,
+    edge_to_stitch_fields: Dict[Tuple[str, str], Tuple[str, str]],
+    name_assigner: IntermediateOutNameAssigner,
+) -> AstType:
     """Return an AST node with which to replace the input AST in the selections that contain it.
 
     This function examines the selections of the input AST, and recursively calls either
-    _split_query_ast_one_level_recursive_type_coercion or
-    _split_query_ast_one_level_recursive_normal_fields
-    depending on whether the selections contains a single InlineFragment or a number of normal
+    _split_query_ast_one_level_recursive or _split_query_ast_one_level_recursive_normal_fields
+    depending on whether the selections contains a single InlineFragmentNode or a number of normal
     fields.
 
     Args:
         query_node: SubQueryNode, whose list of child query connections may be modified to
-                    include new children
-        ast: Field, InlineFragment, or OperationDefinition, the AST that we are trying to split
-             into child components. It is not modified by this function
-        type_info: TypeInfo, used to get information about the types of fields while traversing
-                   the query AST
-        edge_to_stitch_fields: Dict[Tuple(str, str), Tuple(str, str)], mapping
-                               (type name, vertex field name) to
+                    include new children.
+        ast: The AST that we are trying to split into child components.
+             It is not modified by this function.
+        type_info: Used to get information about the types of fields while traversing the query AST.
+        edge_to_stitch_fields: Mapping (type name, vertex field name) to
                                (source field name, sink field name) used in the @stitch directive
-                               for each cross schema edge
-        name_assigner: IntermediateOutNameAssigner, object used to generate and keep track of
-                       names of newly created @output directives
+                               for each cross schema edge.
+        name_assigner: Object used to generate and keep track of names of newly created
+                       @output directives.
 
     Returns:
-        Field, InlineFragment, or OperationDefinition, the AST with which to replace the input
-        AST in the selections that contain it
+        The AST with which to replace the input AST in the selections that contain it.
     """
+    if ast.selection_set is None:
+        raise AssertionError("AST's selection_set cannot be None.")
     type_info.enter(ast.selection_set)
-    selections = ast.selection_set.selections
+    selections = list(ast.selection_set.selections)
```

Comment:
Is this `list()` feasible to avoid as well? It's not terribly expensive on its own, but it doesn't seem like we are mutating `selections` itself, so hopefully it might not be so bad? In the worst case, we might even be able to get away with a cast to the necessary type on the next line instead of a copy of the list.